### PR TITLE
osd: add pg_num_max value

### PR DESCRIPTION
**Pull request by liewegas**
- add the property
- some simple tests
- ensure min < pg_num < max
- make pg_autoscaler respect the max

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/controllers/pool.py [hunk 1]
```diff
@@ -74,6 +74,7 @@
     "fast_read": (bool, ""),
     "options": ({
         "pg_num_min": (int, "")
+        "pg_num_max": (int, "")
```
Comment (neha-ojha) [inline src/pybind/mgr/dashboard/controllers/pool.py:77]: I think this should address https://jenkins.ceph.com/job/ceph-pull-requests/86047/consoleFull#450684626c19247c4-fcb7-4c61-9a5d-7e2b9731c678

```suggestion
        "pg_num_min": (int, ""),
        "pg_num_max": (int, "")
```

**Comment by neha-ojha** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/86047/consoleFull#-108728127277933967-90d1-4877-8d60-89cb08ef4eb1 seems to be another occurrence of https://tracker.ceph.com/issues/53000

**Review by batrick — CHANGES_REQUESTED** [review]
Maybe also update `mgr/volumes` to set a `pg_num_max` on the metadata pool.

File: src/pybind/mgr/mgr_module.py [hunk 2]
```diff
@@ -1008,7 +1008,10 @@ def create_pool(self, pool: str) -> None:
             'pool': pool,
             'pg_num': 1,
             'pg_num_min': 1,
+            'pg_num_max': 32,
         }
+        if pg_num_max:
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/pybind/mgr/mgr_module.py]: oops, that was left over from an earlier draft.  fixed.

**Comment by jdurgin** [comment]
> Maybe also update `mgr/volumes` to set a `pg_num_max` on the metadata pool.

Other metadata pools like rgw's smaller ones (not bucket index) would need this too. We keep coming up with metadata-pool-specific settings - maybe we should have some flag for what is a small metadata pool in general, so we could easily apply new settings there.

**Comment by liewegas** [comment]
> > Maybe also update `mgr/volumes` to set a `pg_num_max` on the metadata pool.
> 
> Other metadata pools like rgw's smaller ones (not bucket index) would need this too. We keep coming up with metadata-pool-specific settings - maybe we should have some flag for what is a small metadata pool in general, so we could easily apply new settings there.

Alternatively, maybe we need something like a "small" flag that indicates that pgs should scale up based on usage (and not down based on capacity)?

**Comment by jdurgin** [comment]
> > > Maybe also update `mgr/volumes` to set a `pg_num_max` on the metadata pool.
> > 
> > 
> > Other metadata pools like rgw's smaller ones (not bucket index) would need this too. We keep coming up with metadata-pool-specific settings - maybe we should have some flag for what is a small metadata pool in general, so we could easily apply new settings there.
> 
> Alternatively, maybe we need something like a "small" flag that indicates that pgs should scale up based on usage (and not down based on capacity)?

It gets a bit complicated mixing the two modes. scale-down relies on the total number of pools to determine how to diivde the pg budget. I suppose it could treat 'small' pools, or pools in scale-up mode, as being processed first, and subtract those pgs from the budget, before dividing the budget evenly among scale-down pools with little usage. What do you think @kamoltat ?

**Comment by liewegas** [comment]
jenkins test this please

**Comment by liewegas** [comment]
Regardless of what we do with pool small or bulk flags, I think we still want a pg_num_max pool property.  I suggest moving forward with this PR.

**Comment by neha-ojha** [comment]
jenkins retest this please

**Comment by neha-ojha** [comment]
> Regardless of what we do with pool small or bulk flags, I think we still want a pg_num_max pool property. I suggest moving forward with this PR.

Agreed!

**Review by neha-ojha — APPROVED** [review]

**Comment by liewegas** [comment]
jenkins test dashboard cephadm

**Comment by liewegas** [comment]
@batrick ok to proceed without updating mgr/volumes?  I'm not sure a max makes sense there.

**Comment by batrick** [comment]
> @batrick ok to proceed without updating mgr/volumes? I'm not sure a max makes sense there.

fine with me!

**Comment by neha-ojha** [comment]
@batrick can you remove your request for changes, if everything looks good to you?
@liewegas have you finished testing this? can we merge it?

**Comment by liewegas** [comment]
jenkins test dashboard cephadm

**Comment by liewegas** [comment]
I think this is okay, but there are still some rados failures that I'm not sure about--going to run it through one more time.

@batrick it looks like the mds upgrade tests are still failing?

**Comment by batrick** [comment]
> I think this is okay, but there are still some rados failures that I'm not sure about--going to run it through one more time.
> 
> @batrick it looks like the mds upgrade tests are still failing?

https://github.com/ceph/ceph/pull/44324

MDS upgrade tests were broken by a recent (RADOS!) PR :)

**Review by batrick — APPROVED** [review]

**Comment by neha-ojha** [comment]
> > I think this is okay, but there are still some rados failures that I'm not sure about--going to run it through one more time.
> > @batrick it looks like the mds upgrade tests are still failing?
> 
> #44324
> 
> MDS upgrade tests were broken by a recent (RADOS!) PR :)

the rados suite isn't affected by https://tracker.ceph.com/issues/53615, I made a change similar to https://github.com/ceph/ceph/pull/44324 to other upgrade tests in https://github.com/ceph/ceph/pull/44161 before we merged https://github.com/ceph/ceph/commit/bd815bd9d6ecdecaab3d2dd9e0f5a18aa795d749, I missed the fs/upgrade tests. FWIW, these upgrade tests were setting the wrong flags :)

Latest rados/cephadm/mds_upgrade_sequence failures can be found in https://pad.ceph.com/p/teuthology_failures

**Comment by liewegas** [comment]
jenkins test make check

**Pull request merged**